csrc/abstract_tensor_schedule.cpp

```diff
@@ -0,0 +1,319 @@
+    for (const AbstractId& abs_id : abstract_.domain) {
+      IterDomain* new_id = replayAbstractId(abs_id);
+      if (new_id == nullptr) {
+        continue;
+      }
+      loop_domain.push_back(new_id);
+    }
+    concrete_->setLoopDomain(loop_domain);
+  }
+
+  ValGroup abstractIdToValGroup(const AbstractId& abs_id) {
+    NVF_ERROR(
+        abs_id.is<ValGroupAndItsGraph>(),
+        "AbstractId must be a ValGroupAndItsGraph");
+    const ValGroupAndItsGraph& vgg = abs_id.as<ValGroupAndItsGraph>();
+    if (graph_ == nullptr) {
+      graph_ = vgg.graph;
+    } else {
+      NVF_ERROR(graph_ == vgg.graph);
+    }
+    return vgg.group;
+  }
+
+  //! Work backward from each loop IterDomain in concrete. When we find an
+  //! IterDomain* with a ValGroup in scheduled_val_groups, we map to it
+  //! (including siblings) in concrete_ids_. Assert that we do find a
+  //! scheduled ValGroup, otherwise we would not know where to place this
+  //! IterDomain.
+  //!
+  //! For example, suppose we have the following:
+  //!
+  //!   concrete_:
+  //!     root: iS0{i0} iS1{i1}
+  //!     loop: iS3{ceilDiv(i0 * i1, 16)} iS4{16}
+  //!     transforms:
+  //!       iS2 = merge(iS0, iS1)
+  //!       iS3, iS4 = split(iS2, 16)
+  //!   graph:
+  //!     ValGroup 0: 0 8
+  //!     ValGroup 1: 1 9
+  //!     ValGroup 2: 2 10
+  //!     ValGroup 3: 3
+  //!     ValGroup 4: 4
+  //!     ValGroup 5: 11
+  //!     ValGroup 6: 12
+  //!     ExprGroup 0: iS11{ceilDiv(i0 * i1, 32)}, iS12{32} = split(iS10, 32)
+  //!   abstract_:
+  //!     {{ValGroup 5, graph}, {ValGroup 6, graph}}
+  //!
+  //! Then findNearestProducers() work backward from the loop domain of
+  //! concrete_ (ValGroups 3 and 4) and will identify ValGroup 2 as a producer
+  //! to abstract_ and map it to the concrete domain iS2{i0}.
+  //!
+  //! Note that any dangling transforms will be discarded; in this example since
+  //! we work from iS2, we will discard the split that produced iS3 and iS4.
+  void findNearestProducers() {
+    // Record all ValGroups that are producers of those in abstract_. This lets
+    // us find starting IterDomains for scheduling concrete's loop domain.
+    // TODO: this should probably use a new more general utility that resembles
+    // StmtSort::getExprsTo in val_graph_visitor.h
+    std::unordered_set<ValGroup> scheduled_val_groups;
+    std::stack<ValGroup> vg_stack;
+    for (const AbstractId& abs_id : abstract_.domain) {
+      vg_stack.push(abstractIdToValGroup(abs_id));
+    }
+    while (!vg_stack.empty()) {
+      ValGroup vg = vg_stack.top();
+      vg_stack.pop();
+      bool inserted = scheduled_val_groups.insert(vg).second;
+      if (!inserted) {
+        // Avoid cycles
+        continue;
+      }
+      for (const ExprGroup& eg : graph_->getDefinitions(vg)) {
+        for (Expr* e : *eg) {
+          for (Val* inp : e->inputs()) {
+            ValGroup vg_inp = graph_->toGroup(inp);
+            vg_stack.push(vg_inp);
+          }
+        }
```

Comment:
I think this is similar to the case below, but does it need to check all exprs in an ExprGroup?

Reply:
You're right. I will convert this to just look at the first `Expr` also.